src/aws/flb_aws_util.c

```diff
@@ -927,3 +932,71 @@ flb_sds_t flb_get_s3_key(const char *format, time_t time, const char *tag,
         }
         return NULL;
 }
+
+/*
+ * Replace %f and %L in time_key_format with milliseconds and nanoseconds. And get the
+ * time format output and output length with strftime().
+ */
+size_t create_time_format_string(char **out_buff_ptr, const char *time_key_format,
+                                 struct flb_time *tms)
+{
+    char millisecond_string[FLB_AWS_MILLISECOND_FORMATTER_LENGTH+1];
+    char nanosecond_string[FLB_AWS_NANOSECOND_FORMATTER_LENGTH+1];
+    char *tmp_buff_time_str;
+    char *mem_for_time_key_format;
+    char *tmp_out_buff;
+    size_t out_size;
+    size_t max_length_for_mem;
+    struct tm time_stamp;
+
+    /* guess the maximum length needed for out_buff, the worst scenario will be only %c in time_key_format,
+    and then %c will be tansfer to 24 chars by strftime, so update length to 12 times to avoid memory overflow. */
+    max_length_for_mem = 12*strlen(time_key_format);
+    mem_for_time_key_format = (char *)flb_calloc(max_length_for_mem, sizeof(char));
+    if (!mem_for_time_key_format) {
+        flb_errno();
+        return -1;
+    }
+    /* use mem_for_time_key_format to store time_key_format */
+    strcpy(mem_for_time_key_format, time_key_format);
+    tmp_buff_time_str = (char *)flb_calloc(max_length_for_mem, sizeof(char));
+    if (!tmp_buff_time_str) {
+        flb_errno();
+        free(mem_for_time_key_format);
+        return -1;
+    }
+
+    tmp_out_buff = (char *)flb_calloc(max_length_for_mem, sizeof(char));
+    if (!tmp_out_buff) {
+        flb_errno();
+        free(mem_for_time_key_format);
+        free(tmp_buff_time_str);
+        return -1;
+    }
+    snprintf(millisecond_string, FLB_AWS_MILLISECOND_FORMATTER_LENGTH+1, "%" PRIu64, (uint64_t) tms->tm.tv_nsec / 1000000);
+    snprintf(nanosecond_string, FLB_AWS_NANOSECOND_FORMATTER_LENGTH+1, "%" PRIu64, (uint64_t) tms->tm.tv_nsec);
+    /* replace %F to millisecond, %L to nanosecond in mem_for_time_key_format */
+    for (int i = 0; i < strlen(mem_for_time_key_format); i++) {
+        if (!strncmp(mem_for_time_key_format+i, FLB_AWS_MILLISECOND_FORMATTER, 2)) {
+            strncat(tmp_buff_time_str, millisecond_string, FLB_AWS_MILLISECOND_FORMATTER_LENGTH+1);
+            i += 1;
+        } else if (!strncmp(mem_for_time_key_format+i, FLB_AWS_NANOSECOND_FORMATTER, 2)) {
+            strncat(tmp_buff_time_str, nanosecond_string, FLB_AWS_NANOSECOND_FORMATTER_LENGTH+1);
+            i += 1;
+        } else {
+            strncat(tmp_buff_time_str,mem_for_time_key_format+i,1);
+        }
+    }
+    memcpy(mem_for_time_key_format, tmp_buff_time_str, strlen(tmp_buff_time_str)+1);
+    gmtime_r(&tms->tm.tv_sec, &time_stamp);
+    out_size = strftime(tmp_out_buff, max_length_for_mem, mem_for_time_key_format, &time_stamp);
```

Comment:
Why don't you just use tmp_buff_time_str. Is mem_for_time_key_format really needed?

Reply:
mem_for_time_key_format is used to replace time_key_format and tmp_buff_time_str is a buffer to buff mem_for_time_key_format. you told me I shouldn't change time_key_format, so I used mem_for_time_key_format instead.

Reply:
You can use time_key_format, it just shouldn't be modified